### PR TITLE
Add core.typed annotations to docs

### PR DESCRIPTION
**Pull request by atroche**
This PR adds basic support for core.typed var annotations.

### CODE REVIEW

File: codox.core/project.clj [hunk 1]
```diff
@@ -4,7 +4,7 @@
   :scm {:dir ".."}
   :license {:name "Eclipse Public License"
             :url "http://www.eclipse.org/legal/epl-v10.html"}
-  :dependencies [[org.clojure/clojure "1.5.1"]
+  :dependencies [[org.clojure/clojure "1.6.0"]
```
Comment (weavejester) [inline codox.core/project.clj]: Why is this change necessary?

File: codox.core/src/codox/reader/clojure.clj [hunk 2]
```diff
@@ -1,11 +1,33 @@
 (ns codox.reader.clojure
   "Read raw documentation information from Clojure source directory."
-  (:import java.util.jar.JarFile)
+  (:import java.util.jar.JarFile
+           (java.io FileNotFoundException))
```
Comment (weavejester) [inline codox.core/src/codox/reader/clojure.clj]: To be consistent with the previous line, prefer `java.io.FileNotFoundException`.

File: codox.core/src/codox/reader/clojure.clj [hunk 3]
```diff
@@ -1,11 +1,33 @@
...
   (:use [codox.utils :only (assoc-some update-some correct-indent)])
   (:require [clojure.java.io :as io]
             [clojure.tools.namespace :as ns]
             [clojure.string :as str]))
 
+(defn depends-on-core-typed? []
+  (try
+    (require '[clojure.core.typed])
```
Comment (weavejester) [inline codox.core/src/codox/reader/clojure.clj]: You can lose the vector here, i.e. `(require 'clojure.core.typed)`

File: codox.core/src/codox/reader/clojure.clj [hunk 4]
```diff
@@ -1,11 +1,33 @@
...
   "Read raw documentation information from Clojure source directory."
-  (:import java.util.jar.JarFile)
+  (:import java.util.jar.JarFile
+           (java.io FileNotFoundException))
   (:use [codox.utils :only (assoc-some update-some correct-indent)])
   (:require [clojure.java.io :as io]
             [clojure.tools.namespace :as ns]
             [clojure.string :as str]))
 
+(defn depends-on-core-typed? []
```
Comment (weavejester) [inline codox.core/src/codox/reader/clojure.clj]: This function has side effects, so perhaps something more like `try-require-core-typed` instead, to indicate that it's not just a pure boolean check.

File: codox.core/src/codox/reader/clojure.clj [hunk 5]
```diff
@@ -1,11 +1,33 @@
...
+      false)))
+
+(defn var->symbol [var]
+  (let [{:keys [ns name]} (meta var)]
+    (symbol (format "%s/%s" (str ns) (str name)))))
+
+; need to use ns-resolve in the following two fns because core.typed
+; gets required dynamically
+(defn typecheck-namespace [namespace]
+  (let [check-ns-fn (ns-resolve 'clojure.core.typed 'check-ns-info)]
```
Comment (weavejester) [inline codox.core/src/codox/reader/clojure.clj]: Maybe bind the function to a symbol of the same name, i.e. `check-ns-info`. The same for `check-form-info` too.

File: codox.core/src/codox/reader/clojure.clj (hunk 4 above)
Comment (weavejester) [inline codox.core/src/codox/reader/clojure.clj]: Actually, I wonder if we could split this into two parts. One that tries to require core.typed, and one that uses a namespace resolve to find whether it has been required in a side-effect free way.

File: codox.core/src/codox/reader/clojure.clj (hunk 3 above)
Comment (weavejester) [inline codox.core/src/codox/reader/clojure.clj]: Also, we should require a namespace that exists in core.typed, not core.typed.rt, otherwise we're going to run into problems. `clojure.core.typed.check` would be a better choice, I think.

File: codox.core/src/codox/reader/clojure.clj [hunk 6]
```diff
@@ -41,12 +63,19 @@
...
+    (if (empty? delayed-errors)
+      (:t ret))))
+
 (defn- read-var [vars var]
   (-> (meta var)
       (select-keys [:name :file :line :arglists :doc :dynamic
                     :added :deprecated :doc/format])
       (update-some :doc correct-indent)
       (assoc-some  :type    (var-type var)
+                   :core.typed/type (when (depends-on-core-typed?)
```
Comment (weavejester) [inline codox.core/src/codox/reader/clojure.clj]: We should prefer `if` when the return value matters and we don't need an implicit `do`.

File: codox.example/project.clj [hunk 7]
```diff
@@ -3,7 +3,7 @@
   :url "http://example.com/FIXME"
   :license {:name "Eclipse Public License"
             :url "http://www.eclipse.org/legal/epl-v10.html"}
-  :dependencies [[org.clojure/clojure "1.3.0"]]
+  :dependencies [[org.clojure/clojure "1.6.0"]]
```
Comment (weavejester) [inline codox.example/project.clj]: We can add typing to the example project, but we need to keep the 1.3.0 dependency to ensure backward compatibility is tested.

Comment (atroche) [inline codox.example/project.clj]: oh, it was because adding `core.typed` as a dependency made the example project (on 1.3.0) fail on `lein doc`. I'll change this back to 1.3.0 and leave `core.typed` out.

**Comment by atroche** [comment]
Thanks for taking the time to help, @weavejester! :shower:

File: codox.core/src/codox/reader/clojure.clj [hunk 8]
```diff
@@ -1,11 +1,36 @@
...
   "Read raw documentation information from Clojure source directory."
-  (:import java.util.jar.JarFile)
+  (:import java.util.jar.JarFile
+           java.io.FileNotFoundException)
   (:use [codox.utils :only (assoc-some update-some correct-indent)])
   (:require [clojure.java.io :as io]
             [clojure.tools.namespace :as ns]
             [clojure.string :as str]))
 
+(defn try-require-core-typed []
```
Comment (weavejester) [inline codox.core/src/codox/reader/clojure.clj]: Maybe make this a little more generic, e.g.

```clojure
(defn try-require [namespace]
  (try
    (require namespace)
    (catch FileNotFoundException _ nil)))
```

File: codox.core/src/codox/reader/clojure.clj [hunk 9]
```diff
@@ -1,11 +1,36 @@
...
+  (try
+    (require 'clojure.core.typed.check)
+    (catch FileNotFoundException _ nil)))
+
+(defn depends-on-core-typed? []
```
Comment (weavejester) [inline codox.core/src/codox/reader/clojure.clj]: I think you could rewrite this as:

```clojure
(defn core-typed? []
  (find-ns 'clojure.core.typed.check))
```

File: codox.core/src/codox/reader/clojure.clj [hunk 10]
```diff
@@ -1,11 +1,36 @@
...
+               (ns-resolve 'clojure.core.typed 'check-form-info)
+               (catch java.lang.Exception _ nil)))))
+
+(defn var->symbol [var]
+  (let [{:keys [ns name]} (meta var)]
+    (symbol (format "%s/%s" (str ns) (str name)))))
+
+; need to use ns-resolve in the following two fns because core.typed
+; gets required dynamically
+(defn typecheck-namespace [namespace]
```
Comment (weavejester) [inline codox.core/src/codox/reader/clojure.clj]: You could maybe rewrite this as:

```clojure
(defn typecheck-namespace [namespace]
  ((find-var 'clojure.core.typed/check-ns-info) namespace))
```

File: codox.core/src/codox/reader/clojure.clj [hunk 11]
```diff
@@ -1,11 +1,36 @@
...
+  (let [check-ns-info (ns-resolve 'clojure.core.typed 'check-ns-info)]
+    (check-ns-info namespace)))
+
+(defn typecheck-var [var]
```
Comment (weavejester) [inline codox.core/src/codox/reader/clojure.clj]: You could also rewrite this as:

```clojure
(defn typecheck-var [var]
  ((find-var 'clojure.core.typed/check-form-info) var))
```

**Comment by weavejester** [comment]
Just added some suggestions about using `find-ns` and `find-var` that should make the code a bit more concise.

**Comment by atroche** [comment]
:shower:

**Comment by weavejester** [comment]
Perfect, thanks. Could you squash down the commits into one? Then I'll check and merge.

**Comment by atroche** [comment]
Try that on for size =)

**Comment by weavejester** [comment]
Thanks. Could you remove the intermediate commit messages from the squashed commit? It just needs the summary line.

**Comment by atroche** [comment]
done =)

File: codox.core/src/codox/writer/html.clj [hunk 12]
```diff
@@ -226,6 +226,9 @@
    [:div.usage
     (for [form (var-usage var)]
       [:code (h (pr-str form))])]
+   (if (:core.typed/type var)
+     [:div.core.typed/type
```
Comment (weavejester) [inline codox.core/src/codox/writer/html.clj]: `:div.core.typed/type` isn't a valid HTML tag name.

**Comment by weavejester** [comment]
I tried this out on the example project, but it's not producing the right results. The type signatures just say:

    (clojure.lang.Var clojure.core.typed/Nothing clojure.core.typed/Any)

You also don't appear to have changed the example project to allow for testing of core.typed. Because HTML pages are somewhat difficult to test automatically, I use the example project as a way of testing that Codox is working correctly.

File: codox.core/src/codox/reader/clojure.clj [hunk 13]
```diff
@@ -1,11 +1,32 @@
...
+  (try
+    (require namespace)
+    (catch FileNotFoundException _ nil)))
+
+(defn core-typed? []
+  (find-ns 'clojure.core.typed.check))
+
+(defn var->symbol [var]
+  (let [{:keys [ns name]} (meta var)]
+    (symbol (format "%s/%s" (str ns) (str name)))))
```
Comment (weavejester) [inline codox.core/src/codox/reader/clojure.clj]: This could be rewritten as:

```clojure
(symbol (str ns) (str name))
```

I also think this funciton needs to be used in `typecheck-var`?

**Comment by weavejester** [comment]
I think during a refactor of `typecheck-var`, the `var->symbol` function got removed and needs to be replaced.

**Comment by oubiwann** [comment]
Heads-up: this PR has been superseded by PR #117.

(Still using the original ticket, though: #80.)

**Comment by weavejester** [comment]
Superseded by #117
